src/audio/mixin_mixout.c

```diff
@@ -885,103 +1086,130 @@ static int mixinout_trigger(struct comp_dev *dev, int cmd)
+		sink_c->stream.channels = md->base_cfg.audio_fmt.channels_count;
 
-		sink_c->buffer_fmt = base_cfg->audio_fmt.interleaving_style;
+		/* Applying channel remapping may produce sink stream with channel count
+		 * different from source channel count.
+		 */
+		sink_id = IPC4_SRC_QUEUE_ID(sink_c->id);
+		if (sink_id < MIXIN_MAX_SINKS) {
+			if (md->sink_config[sink_id].mixer_mode ==
+				IPC4_MIXER_CHANNEL_REMAPPING_MODE)
+				sink_c->stream.channels =
+					md->sink_config[sink_id].output_channel_count;
+		} else {
+			comp_err(dev, "Sink index out of range: %u, max sink count: %u",
+				 (uint32_t)sink_id, MIXIN_MAX_SINKS);
+			buffer_release(sink_c);
+			return -EINVAL;
+		}
 
-		/* 8 ch stream is supported by ch_map and each channel
-		 * is mapped by 4 bits. The first channel will be mapped
-		 * by the bits of 0 ~ 3 and the 2th channel will be mapped
-		 * by bits 4 ~ 7. The N channel will be mapped by bits
-		 * N*4 ~ N*4 + 3
+		/* comp_verify_params() does not modify valid_sample_fmt (a BUG?),
+		 * let's do this here
 		 */
-		for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++)
-			sink_c->chmap[i] = (base_cfg->audio_fmt.ch_map >> i * 4) & 0xf;
+		audio_stream_fmt_conversion(md->base_cfg.audio_fmt.depth,
+					    md->base_cfg.audio_fmt.valid_bit_depth,
+					    &sink_c->stream.frame_fmt,
+					    &sink_c->stream.valid_sample_fmt,
+					    md->base_cfg.audio_fmt.s_type);
 
 		buffer_release(sink_c);
 	}
 
+	/* use BUFF_PARAMS_CHANNELS to skip updating channel count */
 	ret = comp_verify_params(dev, BUFF_PARAMS_CHANNELS, params);
 	if (ret < 0) {
-		comp_err(dev, "mixinout_common_params(): comp_verify_params() failed!");
+		comp_err(dev, "mixin_params(): comp_verify_params() failed!");
 		return -EINVAL;
 	}
 
 	return 0;
 }
 
-static int mixin_params(struct comp_dev *dev,
-			struct sof_ipc_stream_params *params)
-{
-	struct mixin_data *md;
-
-	comp_dbg(dev, "mixin_params()");
-
-	md = comp_get_drvdata(dev);
-	return mixinout_common_params(dev, params, &md->base_cfg);
-}
-
 static int mixout_params(struct comp_dev *dev,
 			 struct sof_ipc_stream_params *params)
 {
 	struct mixout_data *md;
-	int ret;
-	struct comp_buffer *sinkb;
+	struct comp_buffer *sink;
 	struct comp_buffer __sparse_cache *sink_c;
+	int ret;
 	uint32_t sink_period_bytes, sink_stream_size;
 
 	comp_dbg(dev, "mixout_params()");
 
 	md = comp_get_drvdata(dev);
+	base_module_cfg_to_stream_params(&md->base_cfg, params);
 
-	ret = mixinout_common_params(dev, params, &md->base_cfg);
-	if (ret < 0)
-		return ret;
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
+			       source_list);
+	sink_c = buffer_acquire(sink);
+
+	/* comp_verify_params() does not modify valid_sample_fmt (a BUG?), let's do this here */
+	audio_stream_fmt_conversion(md->base_cfg.audio_fmt.depth,
+				    md->base_cfg.audio_fmt.valid_bit_depth,
+				    &sink_c->stream.frame_fmt,
+				    &sink_c->stream.valid_sample_fmt,
+				    md->base_cfg.audio_fmt.s_type);
```

Comment:
is this needed because `comp_verify_params()` needs `sink_c->stream.frame_fmt`?

Reply:
No, comp_verify_params() setups some buffer parameters (like frame_fmt), however, it does not setup valid_sample_fmt. This might be a bug in comp_verify_params(). That is why audio_stream_fmt_conversion() is called to properly set valid_sample_fmt. It also sets frame_fmt, but that is not really necessary as comp_verify_params() will do it once again. To summarize that audio_stream_fmt_conversion() call is a workaround to a missing functionality (a BUG maybe) in comp_verify_params().

Reply:
but then, if `comp_verify_params()` doesn't need it, can this be moved below to only acquire the `sink` buffer once?

Reply:
Done.